lib/typed/coercion/struct_coercer.rb

```diff
@@ -21,7 +21,35 @@ def coerce(type:, value:)
 
         return Failure.new(CoercionError.new("Value of type '#{value.class}' cannot be coerced to #{type} Struct.")) unless value.is_a?(Hash)
 
-        Success.new(T.cast(type, T::Types::Simple).raw_type.from_hash!(HashTransformer.new.deep_stringify_keys(value)))
+        values = {}
+        type.raw_type.props.each do |name, prop|
+          # If the prop is a struct, we need to recursively coerce it
+          if prop[:type_object].try(:raw_type).respond_to?(:props)
+            Typed::HashSerializer
+              .new(schema: Typed::Schema.from_struct(prop[:type_object].raw_type))
+              .deserialize(value[name])
+              .and_then { |struct| Typed::Success.new(values[name] = struct) }
+              .on_error { |error| return Typed::Failure.new(CoercionError.new("Nested hash for #{type} could not be coerced to #{name}, error: #{error}")) }
+          else
+            value = HashTransformer.new.deep_symbolize_keys(value)
+
+            # if the value is nil but the type is nilable, no need to coerce
+            if value[name].nil?
+              if prop[:type_object].try(:valid?, value[name])
+                next
+              else
+                return Typed::Failure.new(CoercionError.new("#{name} is required but nil given"))
+              end
+            end
```

Comment:
@maxveldink so the nil-check already happens in the serializer https://github.com/maxveldink/sorbet-schema/blob/main/lib/typed/serializer.rb#L38-L39 but we still need this code addition because within a struct you also want to make sure non-nilable attributes don't have a nil. Based on our convo earlier I put this here instead of in `coercion` and removed my nil-check in the string coercer. Thoughts?

Reply:
> we still need this code addition because within a struct you also want to make sure non-nilable attributes don't have a nil

Wouldn't we be re-using the `nil` check on a nested struct as well?

Reply:
If you could show me a sample test that would be helpful!

Reply:
no because the nil check is in this method https://github.com/maxveldink/sorbet-schema/blob/main/lib/typed/serializer.rb#L34 but we're not calling that method, we're calling [this](https://github.com/maxveldink/sorbet-schema/blob/main/lib/typed/serializer.rb#L41) directly so we just missed the check

👍 I'll ping you when I've got the tests done